// @langchain/groq/dist/utils/groq-schema.cjs

//#region src/utils/groq-schema.ts
/**
* Make a schema nullable using type array syntax (NOT anyOf).
* Groq strict mode forbids anyOf at top level, so we use type: [T, "null"].
* Handles: simple types, enums, anyOf, arrays, and objects.
* Unknown schema structures (e.g., `$ref`) are returned unchanged.
*/
function makeNullable(schema) {
	if (!schema || typeof schema !== "object") return schema;
	if (schema.type === "null") return schema;
	if (Array.isArray(schema.type) && schema.type.includes("null")) return schema;
	if (Array.isArray(schema.enum)) {
		const nextEnum = schema.enum.includes(null) ? schema.enum : [...schema.enum, null];
		const baseType = Array.isArray(schema.type) ? schema.type.filter((t) => t !== "null") : typeof schema.type === "string" ? [schema.type] : ["string"];
		const nextType = Array.from(new Set([...baseType, "null"]));
		return {
			...schema,
			type: nextType,
			enum: nextEnum
		};
	}
	if (Array.isArray(schema.anyOf)) {
		if (schema.anyOf.some((s) => s?.type === "null" || Array.isArray(s?.type) && s.type.includes("null"))) return schema;
		const types = schema.anyOf.map((s) => s?.type).filter((t) => t && typeof t === "string");
		if (types.length === schema.anyOf.length) {
			const { anyOf: _, ...rest } = schema;
			return {
				...rest,
				type: [...types, "null"]
			};
		}
		return {
			...schema,
			anyOf: [...schema.anyOf, { type: "null" }]
		};
	}
	if (typeof schema.type === "string") return {
		...schema,
		type: [schema.type, "null"]
	};
	if (Array.isArray(schema.type)) return {
		...schema,
		type: Array.from(new Set([...schema.type, "null"]))
	};
	if (schema.properties) return {
		...schema,
		type: ["object", "null"]
	};
	return schema;
}
/**
* Transform a JSON Schema to be Groq strict-mode compatible.
*
* Recursively:
* 1. Sets additionalProperties: false on all objects
* 2. Moves all properties to required array
* 3. Makes previously optional properties nullable
* 4. Throws if root schema has oneOf/not; extracts object variant from anyOf (Groq forbids these at top level)
*
* @param schema - The schema to transform
* @param isRoot - Whether this is the root schema (default: true on first call)
*/
function groqStrictifySchema(schema, isRoot = true) {
	if (!schema || typeof schema !== "object") return schema;
	if (Array.isArray(schema)) return schema.map((s) => groqStrictifySchema(s, false));
	const result = { ...schema };
	for (const k of [
		"anyOf",
		"oneOf",
		"allOf"
	]) if (Array.isArray(result[k])) result[k] = result[k].map((s) => groqStrictifySchema(s, false));
	if (result.items) result.items = groqStrictifySchema(result.items, false);
	if (result.$defs) {
		result.$defs = { ...result.$defs };
		for (const key of Object.keys(result.$defs)) result.$defs[key] = groqStrictifySchema(result.$defs[key], false);
	}
	if ((result.type === "object" || Array.isArray(result.type) && result.type.includes("object") || result.properties) && result.properties && typeof result.properties === "object") {
		result.additionalProperties = false;
		const propKeys = Object.keys(result.properties);
		const prevRequired = new Set(Array.isArray(result.required) ? result.required : []);
		result.properties = { ...result.properties };
		for (const key of propKeys) {
			result.properties[key] = groqStrictifySchema(result.properties[key], false);
			if (!prevRequired.has(key)) result.properties[key] = makeNullable(result.properties[key]);
		}
		result.required = propKeys;
	}
	if (isRoot) {
		if (Array.isArray(result.anyOf)) {
			const objectVariant = result.anyOf.find((s) => s?.type === "object" || s?.properties);
			if (objectVariant) return {
				...groqStrictifySchema(objectVariant, true),
				type: "object"
			};
			throw new Error("Groq strict mode requires the root schema to be an object type, but the schema has anyOf/oneOf with no object variant.");
		}
		if (result.oneOf) throw new Error("Groq strict mode does not support oneOf at the root schema level.");
		if (result.not) throw new Error("Groq strict mode does not support 'not' at the root schema level.");
		if (result.properties) result.type = "object";
	}
	return result;
}
/**
* Check if a model supports native JSON Schema structured output.
* Uses prefix matching so new gpt-oss models are automatically supported.
*/
function supportsJsonSchema(model) {
	return model.startsWith("openai/gpt-oss");
}
/**
* Supported structured output methods for Groq.
*/
const SUPPORTED_STRUCTURED_OUTPUT_METHODS = [
	"jsonSchema",
	"functionCalling",
	"jsonMode"
];
/**
* Get the structured output method for a given Groq model.
*
* - Returns `jsonSchema` if the model supports native JSON Schema and no method is specified
* - Throws if an invalid method is provided
* - Throws if `jsonSchema` is requested for a model that doesn't support it
*
* @param model - The model name
* @param method - Optional method override
* @returns The structured output method to use
*/
function getGroqStructuredOutputMethod(model, method) {
	if (method !== void 0 && !SUPPORTED_STRUCTURED_OUTPUT_METHODS.includes(method)) throw new Error(`Invalid structured output method: ${method}. Supported methods are: ${SUPPORTED_STRUCTURED_OUTPUT_METHODS.join(", ")}`);
	const modelSupportsJsonSchema = supportsJsonSchema(model);
	if (modelSupportsJsonSchema && !method) return "jsonSchema";
	if (!modelSupportsJsonSchema && method === "jsonSchema") throw new Error(`Native JSON Schema structured output is not supported for model "${model}". Only models with the "openai/gpt-oss" prefix are supported. Use "functionCalling" or "jsonMode" instead.`);
	return method ?? "functionCalling";
}

//#endregion
exports.getGroqStructuredOutputMethod = getGroqStructuredOutputMethod;
exports.groqStrictifySchema = groqStrictifySchema;
